pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+		/// migrations.
+		fn on_runtime_upgrade() -> Weight {
+			log::warn!("Performing on_runtime_upgrade");
+
+			let mut weight: Weight = 0u64.into();
+			// TODO: derive a suitable value here, which is probably something < max_block
+			let available_weight: Weight = T::BlockWeights::get().max_block;
+
+			// start by flagging that we are not fully upgraded
+			<FullyUpgraded<T>>::put(false);
+			weight += T::DbWeight::get().writes(1);
+			Self::deposit_event(Event::RuntimeUpgradeStarted());
+
+			weight += perform_runtime_upgrades::<T>(available_weight.saturating_sub(weight));
+
+			if !<FullyUpgraded<T>>::get() {
+				log::error!(
+					"migrations weren't completed in on_runtime_upgrade(), but we're not
+				configured for multi-block migrations; state is potentially inconsistent!"
+				);
+			}
+
+			weight
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_fully_upgraded)]
+	/// True if all required migrations have completed
+	type FullyUpgraded<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn migration_state)]
+	/// MigrationState tracks the progress of a migration.
+	/// Maps name (Vec<u8>) -> whether or not migration has been completed (bool)
+	type MigrationState<T: Config> = StorageMap<_, Twox64Concat, Vec<u8>, bool, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub completed_migrations: Vec<Vec<u8>>,
+		pub phantom: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				completed_migrations: vec![],
+				phantom: PhantomData,
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for migration_name in &self.completed_migrations {
+				<MigrationState<T>>::insert(migration_name, true);
+			}
+		}
+	}
+
+	fn perform_runtime_upgrades<T: Config>(available_weight: Weight) -> Weight {
+		let mut weight: Weight = 0u64.into();
+
+		for migration in &T::MigrationsList::get() {
+			let migration_name = migration.friendly_name();
+			let migration_name_as_bytes = migration_name.as_bytes();
+			log::trace!("evaluating migration {}", migration_name);
+
+			let migration_done = <MigrationState<T>>::get(migration_name_as_bytes).unwrap_or(false);
+
+			if !migration_done {
+				<Pallet<T>>::deposit_event(Event::MigrationStarted(migration_name_as_bytes.into()));
+
+				// when we go overweight, leave a warning... there's nothing we can really do about
+				// this scenario other than hope that the block is actually accepted.
+				let available_for_step = if available_weight > weight {
+					available_weight - weight
+				} else {
+					log::error!(
+						"previous migration went overweight;
+						ignoring and providing migration {} 0 weight.",
+						migration_name,
+					);
+
+					0u64.into()
+				};
+
+				log::trace!(
+					"performing migration {}, avail weight: {}",
```

Comment:
```suggestion
					"performing migration {}, available weight: {}",
```
I just never saw this abbreviation before